File: r_pi_func_dir/pi.r
#!/usr/bin/env Rscript

# This implementation uses a custom function on a vector rather than a 
# loop which is kinda neat, but can be memory consuming.

numslices <- 100000000

args <- commandArgs(trailingOnly=TRUE)

if (length(args)>0) {
  numslices <- strtoi(args)
}

cat("Calculating PI using", numslices, "slices\n")
cat("  1 process\n")

start <- Sys.time()

# Define our function
y <- function(x) {
  1/(1+(x^2))
}

# Sum our function applied to the slices.
mypi <- (4 * sum(y((1.0/numslices) * 1:numslices)))/numslices

stop <- Sys.time()
elapsed <- stop - start

cat("Obtained value of PI:", mypi, "\n")
cat("Time taken:", elapsed, "seconds\n")

File: python_functional_pi_dir/pi.py
def y(x):
    return 1.0/(1.0 + (x**2))

def calcpi(n, r):
    return (4.0/n) * sum(map(y, r))

# Use a generator to save memory (lists are huge!).
def gen_x(n):
    for i in range(n):
        yield (i + 0.5)/n

# In lieu of a main.
if __name__=='__main__':
    import sys
    import time

    num_steps = 5000000

    if len(sys.argv) > 1:
        num_steps = int(sys.argv[1])

    print('Calculating PI with:\n  ' + str(num_steps) + ' slices')
    start =  time.time()
    gen = gen_x(num_steps)
    p = calcpi(num_steps, gen)
    t = time.time() - start
    print('Obtained value of PI: ' + str(p))
    print('Time taken:           ' + str(t) + ' seconds')